### ISSUE TITLE
Books crashing client

### ISSUE THREAD

**Issue by xana43** [PaperMC/Paper#3507]
### Link/paste of stack trace

---- Minecraft Crash Report ----
// Would you like a cupcake?

Time: 6/5/20 9:04 AM
Description: Rendering screen

java.lang.StringIndexOutOfBoundsException: String index out of range: 254
	at java.lang.String.substring(String.java:1951)
	at dha.m(SourceFile:573)
	at dha.render(SourceFile:532)
	at dqn.a(SourceFile:604)
	at dbn.d(SourceFile:989)
	at dbn.d(SourceFile:619)
	at net.minecraft.client.main.Main.main(SourceFile:204)

A detailed walkthrough of the error, its code path and all known details is as follows:
---------------------------------------------------------------------------------------

-- Head --
Thread: Render thread
Stacktrace:
	at java.lang.String.substring(String.java:1951)
	at dha.m(SourceFile:573)
	at dha.render(SourceFile:532)

-- Screen render details --
Details:
	Screen name: dha
	Mouse location: Scaled: (266, 167). Absolute: (798.000000, 502.000000)
	Screen size: Scaled: (617, 353). Absolute: (1851, 1057). Scale factor of 3.000000

-- Affected level --
Details:
	All players: 1 total; [dqb['Ultra_Nerd'/56476, l='MpServer', x=1166.33, y=72.00, z=3320.15]]
	Chunk stats: Client Chunk Cache: 5041, 4225
	Level dimension: minecraft:overworld
	Level name: MpServer
	Level seed: -2074543133100514335
	Level generator: ID 00 - default, ver 1. Features enabled: false
	Level generator options: {}
	Level spawn location: World: (5696,111,4178), Chunk: (at 0,6,2 in 356,261; contains blocks 5696,0,4176 to 5711,255,4191), Region: (11,8; contains chunks 352,256 to 383,287, blocks 5632,0,4096 to 6143,255,4607)
	Level time: 126551 game time, 2142683042 day time
	Known server brands: 
	Level was modded: false
	Level storage version: 0x00000 - Unknown?
	Level weather: Rain time: 0 (now: false), thunder time: 0 (now: false)
	Level game mode: Game mode: survival (ID 0). Hardcore: false. Cheats: false
	Server brand: Paper
	Server type: Non-integrated multiplayer server
Stacktrace:
	at dno.a(SourceFile:450)
	at dbn.c(SourceFile:1921)
	at dbn.d(SourceFile:634)
	at net.minecraft.client.main.Main.main(SourceFile:204)

-- System Details --
Details:
	Minecraft Version: 1.15.2
	Minecraft Version ID: 1.15.2
	Operating System: Windows 10 (amd64) version 10.0
	Java Version: 1.8.0_51, Oracle Corporation
	Java VM Version: Java HotSpot(TM) 64-Bit Server VM (mixed mode), Oracle Corporation
	Memory: 4949728584 bytes (4720 MB) / 8942256128 bytes (8528 MB) up to 17179869184 bytes (16384 MB)
	CPUs: 16
	JVM Flags: 9 total; -XX:HeapDumpPath=MojangTricksIntelDriversForPerformance_javaw.exe_minecraft.exe.heapdump -Xss1M -Xmx16G -XX:+UnlockExperimentalVMOptions -XX:+UseG1GC -XX:G1NewSizePercent=20 -XX:G1ReservePercent=20 -XX:MaxGCPauseMillis=50 -XX:G1HeapRegionSize=16M
	Launched Version: 1.15.2
	Backend library: LWJGL version 3.2.2 build 10
	Backend API: GeForce GTX 1080/PCIe/SSE2 GL version 4.6.0 NVIDIA 446.14, NVIDIA Corporation
	GL Caps: Using framebuffer using OpenGL 3.0
	Using VBOs: Yes
	Is Modded: Probably not. Jar signature remains and client brand is untouched.
	Type: Client (map_client.txt)
	Resource Packs: vanilla
	Current Language: English (US)
	CPU: 16x AMD Ryzen 7 3800X 8-Core Processor 

### Plugin list:
AdminOnly,AsyncWorldEdit,AutoPluginLoader,BetterShulkerBoxes,BetterSleeping3,BKCommonLib,ChatControl,ClearLag,ColoredBooked,CoreProtect,CraftBook,CS-CoreLib,CustomCommands,dynmap,Dynmap-Towny,EntityTrackerFixer,EssentialsX,EssentialsXAntiBuild,EssentialsXChat,EssentialsProtect,EssentialsSpawn,Jobs,LitePlaytimeRewards,LockettePro,LuckPerms,MCJukebox,mcMMO,MorphRedeem,Movecraft,Multiverse-Core,Multiverse-Inventories,Multiverse-Portals,Oh_the_dungeons_youll_go,PlaceholderAPI,ProtocolLib,RandomTeleport,SafeTeleport,ServerRestorer,Slimefun,SuperVanish,TAB,Towny,TownyChat,TownyNameUpdater,UltimateAutoRestart,Vault,ViaBackwards,ViaRewind,ViaVersion,vivecraft-Spigot-Extensions,WG-GUI,WitherAC,WorldEdit,WorldGuard,WorldGuard-Towny,WorldGuardExtraFlags,Factions,FongusCore,GCore,GoldenEnchants,GriefPrevention.

### Actions to reproduce (if known):
I'm not sure if it's an issue with how paper handles colorcodes or if it's an issue with ColoredBooked, but in order to reproduce the issue, make a book with the minecraft text color formatting codes, if they're used enough the client will crash out without leaving a trace on the server

### Paper build number:
Paper 334

**Comment by xana43** [comment]
also it seems that extra texts that are present in the books are cut off, and when viewed in their writing form the NBT data for the text is still there, and the text is readable but that appears to the form of the book that crashes the game

**Comment by Black-Hole** [comment]
This is a client issue. I don't know if the server should filter all book data that might crash a client.
https://bugs.mojang.com/browse/MC-177788

**Comment by aikar** [comment]
Is this a creative server? If so, fixing this is the responsibility of the server owner. Creative mode is dangerous and it should not be given to untrusted users.

Only an issue for Paper is if this book was created in Survival.

**Comment by xana43** [comment]
this server is a survival server, and this book was in fact written in survival. Also if this helps, the issue went away when I downgraded to paper 331

**Comment by aikar** [comment]
We can not do anything about a client bug.
The irony is, the build you said 'works' has all the same changes as latest build too...

I verified a book sent to the client that caused me to crash looks perfectly valid:
```
[03:35:20 INFO]: Item: {pages:["&bFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBar&vc
[03:35:20 INFO]: 
[03:35:20 INFO]: &cFooFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBar&vc
[03:35:20 INFO]: 
[03:35:20 INFO]: &cFoo","&bFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBar&vc
[03:35:20 INFO]: 
[03:35:20 INFO]: &cFooFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBarFooBar&vc
[03:35:20 INFO]: 
[03:35:20 INFO]: &cFoo","&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo","&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo","&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo","&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo","&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo","&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo","&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo&cwef&afoo"]}
```

We can't do anything about that.

**Comment by aikar** [comment]
nvm. I prob typed that in spam. 

Even
[03:57:41 INFO]: Item: {pages:["&cdd&cd&bBar
[03:57:41 INFO]: &0","","","dddd"]}

did it.. So yeah client is just broken.

Might deal with ending in a color code.

**Comment by aikar** [comment]
My suggestion would be for the plugin author to trim any trailing color codes on new lines with no text after it if it's the last part of the page

**Comment by aikar** [comment]
@ForceUpdate1

**Comment by kennytv** [comment]
at least it's already fixed for the client https://bugs.mojang.com/browse/MC-149792

**Comment by Black-Hole** [comment]
The stack traces for https://bugs.mojang.com/browse/MC-177788 and https://bugs.mojang.com/browse/MC-149792 are looking different. So this might be fixed in 1.16 or not.

**Comment by kennytv** [comment]
it was resolved in the same version, my issue contained that trace as well and was proven to be fixed (see attachements)
